src/backend/pipeline/cont_combiner.c

```diff
@@ -812,9 +786,18 @@ read_batch(ContQueryCombinerState *state, TupleBufferBatchReader *reader)
 
 		ExecStoreTuple(heap_copytuple(tbs->tuple->heaptup), state->slot, InvalidBuffer, false);
 		tuplestore_puttupleslot(state->batch, state->slot);
+		state->batch_hashes[count] = tbs->tuple->group_hash;
 
 		IncrementCQRead(1, tbs->size);
 		count++;
+
+		/*
+		 * This is fairly unlikely, but if it does happen we should
+		 * finish processing this batch immediately because we're storing 8-byte hashes
+		 * in memory alongside each of these tuples.
+		 */
+		if (!tuplestore_in_memory(state->batch))
```

Comment:
No, keep it. I'm now wondering if we should stop when memory is <= half of 256mb (or whatever the combiner working memory is), because in worst case we could look up exactly the number of tuples from the matrel as there exist in the batch?

Moreover, the size of the tuple buffer is only 64mb which seems too low compared to the 256mb working memory for the batch being read from it?